src/Symfony/Component/HttpClient/ResponseSetMonitor.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Contracts\HttpClient\Exception\ExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Monitors a set of responses and triggers callbacks as they complete.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ResponseSetMonitor implements \Countable
+{
+    private $client;
+    private $onHeaders;
+    private $onBody;
+    private $onError;
+    private $responses;
+
+    public function __construct(HttpClientInterface $client, callable $onHeaders = null, callable $onBody = null, callable $onError = null)
```

Comment:
IMHO, the onHeaders callable should be the last argument, it's less useful than the body or error.
Same for `add` method